contrib/linux/sensors/file_watch_sensor.py

```diff
@@ -14,56 +14,218 @@
 # limitations under the License.
 
 import os
+import signal
+import time
+import sys
 
 import eventlet
+from watchdog.observers import Observer
+from watchdog.events import FileSystemEventHandler
 
-from logshipper.tail import Tail
+try:
+    from st2reactor.sensor.base import Sensor
+except ImportError:
+    Sensor = object
 
-from st2reactor.sensor.base import Sensor
+
+class FileEventHandler(FileSystemEventHandler):
+    def __init__(self, *args, callbacks=None, **kwargs):
+        self.callbacks = callbacks or {}
+
+    def dispatch(self, event):
+        if not event.is_synthetic and not event.is_directory:
+            super().dispatch(event)
+
+    def on_created(self, event):
+        cb = self.callbacks.get('created')
+        if cb:
+            cb(event=event)
+
+    def on_modified(self, event):
+        cb = self.callbacks.get('modified')
+        if cb:
+            cb(event=event)
+
+    def on_moved(self, event):
+        cb = self.callbacks.get('moved')
+        if cb:
+            cb(event=event)
+
+    def on_deleted(self, event):
+        cb = self.callbacks.get('deleted')
+        if cb:
+            cb(event=event)
+
+
+class SingleFileTail(object):
+    def __init__(self, path, handler, read_all=False, observer=None):
+        self.path = path
+        self.handler = handler
+        self.read_all = read_all
+        self.buffer = ''
+        self.observer = observer or Observer()
+
+        self.open()
+
+    def read(self, event=None):
+        while True:
+            # Buffer 1024 bytes at a time
+            buff = os.read(self.fd, 1024)
+            if not buff:
+                return
+
+            # Possible bug? What if the 1024 cuts off in the middle of a utf8
+            # code point?
+            # We use errors='replace' to have Python replace the unreadable
+            # character with an "official U+FFFD REPLACEMENT CHARACTER"
+            # This isn't great, but it's better than the previous behavior,
+            # which blew up on any issues.
+            buff = buff.decode(encoding='utf8', errors='replace')
```

Comment:
We encountered similar issues in Libcloud - another option would be to only decode the buffer once we have the whole line (we do something similar here - https://github.com/apache/libcloud/blob/trunk/libcloud/compute/ssh.py#L601).

Reply:
I fixed this and added 30+ tests for this case. The code will now try to calculate how many more bytes to read, read exactly that many, and try decoding again.